## 5. 🔎 Análisis Exploratorio de Datos

### 5.1 Carga de librerías y datos

En esta sección se realiza la carga de las librerías necesarias para la exploración, análisis y visualización del conjunto de datos. Posteriormente, se importa el archivo `Data_set.csv`, que contiene la información de las unidades analizadas y las variables relacionadas con Riesgo Psicosocial, Evaluación del Desempeño y Engagement.

In [1]:
# Librerías para manipulación y análisis de datos
import pandas as pd
import numpy as np

# Librerías para visualización
import matplotlib.pyplot as plt
import seaborn as sns

# Configuración para visualizar todas las columnas
pd.set_option('display.max_columns', None)

# Carga del conjunto de datos
df = pd.read_csv('../data/Data_set.csv', delimiter=';')

### 5.2 Exploración y calidad de los datos

In [2]:
# Información general del conjunto de datos
df.info()

# Visualización de los primeros registros
df.head()

<class 'pandas.DataFrame'>
RangeIndex: 273 entries, 0 to 272
Data columns (total 61 columns):
 #   Column                                                      Non-Null Count  Dtype
---  ------                                                      --------------  -----
 0   REGIONAL                                                    273 non-null    str  
 1   CODIGO UNIDAD                                               273 non-null    str  
 2   TIPO                                                        273 non-null    str  
 3   APERTURA                                                    273 non-null    int64
 4   Riesgo psicosocial_AÑO 2025_ESTRÉS_Riesgo muy alto          221 non-null    str  
 5   Riesgo psicosocial_AÑO 2025_ESTRÉS_Riesgo alto              221 non-null    str  
 6   Riesgo psicosocial_AÑO 2025_ESTRÉS_Riesgo medio             221 non-null    str  
 7   Riesgo psicosocial_AÑO 2025_ESTRÉS_Riesgo bajo              221 non-null    str  
 8   Riesgo psicosocial_AÑO 2025_EST

,REGIONAL,CODIGO UNIDAD,TIPO,APERTURA,Riesgo psicosocial_AÑO 2025_ESTRÉS_Riesgo muy alto,Riesgo psicosocial_AÑO 2025_ESTRÉS_Riesgo alto,Riesgo psicosocial_AÑO 2025_ESTRÉS_Riesgo medio,Riesgo psicosocial_AÑO 2025_ESTRÉS_Riesgo bajo,Riesgo psicosocial_AÑO 2025_ESTRÉS_Sin riesgo,Riesgo psicosocial_AÑO 2025_INTRALABORALES_Riesgo muy alto,Riesgo psicosocial_AÑO 2025_INTRALABORALES_Riesgo alto,Riesgo psicosocial_AÑO 2025_INTRALABORALES_Riesgo medio,Riesgo psicosocial_AÑO 2025_INTRALABORALES_Riesgo bajo,Riesgo psicosocial_AÑO 2025_INTRALABORALES_Sin riesgo,Riesgo psicosocial_AÑO 2025_EXTRABORALES_Riesgo muy alto,Riesgo psicosocial_AÑO 2025_EXTRABORALES_Riesgo alto,Riesgo psicosocial_AÑO 2025_EXTRABORALES_Riesgo medio,Riesgo psicosocial_AÑO 2025_EXTRABORALES_Riesgo bajo,Riesgo psicosocial_AÑO 2025_EXTRABORALES_Sin riesgo,Riesgo psicosocial_AÑO 2023_ESTRÉS_Riesgo muy alto,Riesgo psicosocial_AÑO 2023_ESTRÉS_Riesgo alto,Riesgo psicosocial_AÑO 2023_ESTRÉS_Riesgo medio,Riesgo psicosocial_AÑO 2023_ESTRÉS_Riesgo bajo,Riesgo psicosocial_AÑO 2023_ESTRÉS_Sin riesgo,Riesgo psicosocial_AÑO 2023_INTRALABORALES_Riesgo muy alto,Riesgo psicosocial_AÑO 2023_INTRALABORALES_Riesgo alto,Riesgo psicosocial_AÑO 2023_INTRALABORALES_Riesgo medio,Riesgo psicosocial_AÑO 2023_INTRALABORALES_Riesgo bajo,Riesgo psicosocial_AÑO 2023_INTRALABORALES_Sin riesgo,Riesgo psicosocial_AÑO 2023_EXTRABORALES_Riesgo muy alto,Riesgo psicosocial_AÑO 2023_EXTRABORALES_Riesgo alto,Riesgo psicosocial_AÑO 2023_EXTRABORALES_Riesgo medio,Riesgo psicosocial_AÑO 2023_EXTRABORALES_Riesgo bajo,Riesgo psicosocial_AÑO 2023_EXTRABORALES_Sin riesgo,Evaluación del desempeño_AÑO 2025_OBJETIVO,Evaluación del desempeño_AÑO 2025_COMPETENCIA,Evaluación del desempeño_AÑO 2025_RESULTADO,Evaluación del desempeño_AÑO 2025_INTERPRETACION,Evaluación del desempeño_AÑO 2024_OBJETIVO,Evaluación del desempeño_AÑO 2024_COMPETENCIA,Evaluación del desempeño_AÑO 2024_RESULTADO,Evaluación del desempeño_AÑO 2024_INTERPRETACION,Evaluación del desempeño_AÑO 2023_OBJETIVO,Evaluación del desempeño_AÑO 2023_COMPETENCIA,Evaluación del desempeño_AÑO 2023_RESULTADO,Evaluación del desempeño_AÑO 2023_INTERPRETACION,Evaluación del desempeño_AÑO 2022_OBJETIVO,Evaluación del desempeño_AÑO 2022_COMPETENCIA,Evaluación del desempeño_AÑO 2022_RESULTADO,Evaluación del desempeño_AÑO 2022_INTERPRETACION,Evaluación del desempeño_AÑO 2021_OBJETIVO,Evaluación del desempeño_AÑO 2021_COMPETENCIA,Evaluación del desempeño_AÑO 2021_RESULTADO,Evaluación del desempeño_AÑO 2021_INTERPRETACION,Evaluación del desempeño_AÑO 2020_OBJETIVO,Evaluación del desempeño_AÑO 2020_COMPETENCIA,Evaluación del desempeño_AÑO 2020_RESULTADO,Evaluación del desempeño_AÑO 2020_INTERPRETACION,Engagement_AÑO 2025_AMBIENTE LABORAL,Engagement_AÑO 2025_COMPROMISO,Engagement_AÑO 2025_INTERPRETACION
0,CAUCA,A4112,Regional,42036,"0,315589354","0,247148289","0,216730038","0,106463878","0,114068441","0,133079848","0,201520913","0,258555133","0,220532319","0,186311787","0,197718631","0,212927757","0,25095057","0,140684411","0,197718631","0,162629758","0,186851211","0,242214533","0,155709343","0,252595156","0,093425606","0,173010381","0,211072664","0,249134948","0,273356401","0,162629758","0,186851211","0,242214533","0,155709343","0,252595156","0,897255455","0,885714286","0,893793104",Alto,"0,6435","0,838055556","0,70782",Medio,"0,699578571","0,844571429","0,743076429",Medio,"0,837","0,8668","0,84594",Alto,"0,90875","0,882986111","0,895868056",Alto,"0,8515","0,859861111","0,855680556",Alto,"0,963690476",1,Muy alto
1,CAUCA,A411243,PDA,39539,"0,222222222","0,222222222","0,333333333","0,222222222",0,"0,222222222","0,222222222","0,333333333","0,111111111","0,111111111","0,111111111","0,222222222","0,555555556","0,111111111",0,"0,363636364","0,181818182",0,"0,272727273","0,181818182","0,090909091","0,090909091","0,181818182","0,454545455","0,181818182","0,181818182","0,090909091","0,363636364","0,363636364",0,"0,807467273","0,847272727","0,819408909",Alto,"0,623636364","0,8391414

In [3]:
# Validar el numero de filas y columnas del DataFrame
df.shape

(273, 61)

#### 5.2.1 Validación de valores faltantes y selección de unidades de análisis

Se realiza una revisión inicial de los valores faltantes del conjunto de datos para identificar posibles limitaciones en la información disponible. Para garantizar la comparabilidad entre las unidades analizadas y considerando la disponibilidad de información en las dimensiones de Riesgo Psicosocial, Desempeño y Engagement, el análisis se concentrará en las unidades de tipo **Agencia** y **PDA**. Las unidades correspondientes a **Dirección General** y **Regional** se excluyen del análisis de clusterización.

In [4]:
# Distribución de unidades por tipo
df['TIPO'].value_counts()

TIPO
Agencia              121
Dirección general     74
PDA                   65
Regional              13
Name: count, dtype: int64

In [5]:
# Validación de valores faltantes
resumen_nulos = pd.DataFrame({
    'Valores faltantes': df.isna().sum(),
    'Porcentaje (%)': (df.isna().mean() * 100).round(2)
})

resumen_nulos[resumen_nulos['Valores faltantes'] > 0].sort_values(
    'Porcentaje (%)',
    ascending=False
)

,Valores faltantes,Porcentaje (%)
Riesgo psicosocial_AÑO 2023_EXTRABORALES_Riesgo alto,67,24.54
Riesgo psicosocial_AÑO 2023_ESTRÉS_Riesgo alto,67,24.54
Riesgo psicosocial_AÑO 2023_EXTRABORALES_Sin riesgo,67,24.54
Riesgo psicosocial_AÑO 2023_EXTRABORALES_Riesgo bajo,67,24.54
Riesgo psicosocial_AÑO 2023_EXTRABORALES_Riesgo medio,67,24.54
Riesgo psicosocial_AÑO 2023_EXTRABORALES_Riesgo muy alto,67,24.54
Riesgo psicosocial_AÑO 2023_INTRALABORALES_Sin riesgo,67,24.54
Riesgo psicosocial_AÑO 2023_INTRALABORALES_Riesgo bajo,67,24.54
Riesgo psicosocial_AÑO 2023_INTRALABORALES_Riesgo medio,67,24.54
Riesgo psicosocial_AÑO 2023_INTRALABORALES_Riesgo alto,67,24.54


In [6]:
# Seleccionar únicamente Agencias y PDA
df_modelo = df[df['TIPO'].isin(['Agencia', 'PDA'])].copy()

# Verificar dimensiones
print(f"Registros iniciales: {len(df)}")
print(f"Registros para el análisis: {len(df_modelo)}")

print("\nDistribución de unidades:")
print(df_modelo['TIPO'].value_counts())

print ("----" * 20)
# Verificar tipos incluidos en el dataset de análisis
df_modelo['TIPO'].unique()

Registros iniciales: 273
Registros para el análisis: 186

Distribución de unidades:
TIPO
Agencia    121
PDA         65
Name: count, dtype: int64
--------------------------------------------------------------------------------


<StringArray>
['PDA', 'Agencia']
Length: 2, dtype: str

### 5.3 Transformación y ajuste de tipos de datos

Durante la exploración inicial se identificó que varias variables cuantitativas fueron cargadas como texto (`string`), a pesar de contener valores numéricos. Esto se debe principalmente al formato utilizado para representar los decimales. También se identificó que la variable `APERTURA` corresponde a una fecha almacenada como número serial de Excel. Se realizará la conversión de estas variables a sus tipos de datos correspondientes, manteniendo como variables categóricas aquellas que representan identificadores, tipos de unidad, regionales e interpretaciones.

In [14]:
# Variables que deben permanecer como categóricas
variables_categoricas = [
    'REGIONAL',
    'CODIGO UNIDAD',
    'TIPO'
]

# Variables de interpretación
variables_interpretacion = [
    col for col in df.columns 
    if 'INTERPRETACION' in col
]

variables_categoricas = variables_categoricas + variables_interpretacion

# Variables candidatas a conversión numérica
variables_numericas = [
    col for col in df.columns 
    if col not in variables_categoricas + ['APERTURA']
]

len(variables_numericas)

# Convertir variables numéricas de texto a formato numérico
for col in variables_numericas:
    df_modelo[col] = pd.to_numeric(
        df[col].astype(str).str.replace(',', '.', regex=False),
        errors='coerce'
    )

# Convertir APERTURA de número serial de Excel a fecha
df_modelo['APERTURA'] = pd.to_datetime(
    df_modelo['APERTURA'],
    unit='D',
    origin='1899-12-30'
)
    
# Información general del conjunto de datos
df_modelo.info()

# Visualización de los primeros registros
df_modelo.head()

<class 'pandas.DataFrame'>
Index: 186 entries, 1 to 272
Data columns (total 61 columns):
 #   Column                                                      Non-Null Count  Dtype        
---  ------                                                      --------------  -----        
 0   REGIONAL                                                    186 non-null    str          
 1   CODIGO UNIDAD                                               186 non-null    str          
 2   TIPO                                                        186 non-null    str          
 3   APERTURA                                                    186 non-null    datetime64[s]
 4   Riesgo psicosocial_AÑO 2025_ESTRÉS_Riesgo muy alto          186 non-null    float64      
 5   Riesgo psicosocial_AÑO 2025_ESTRÉS_Riesgo alto              186 non-null    float64      
 6   Riesgo psicosocial_AÑO 2025_ESTRÉS_Riesgo medio             186 non-null    float64      
 7   Riesgo psicosocial_AÑO 2025_ESTRÉS_Riesgo bajo     

,REGIONAL,CODIGO UNIDAD,TIPO,APERTURA,Riesgo psicosocial_AÑO 2025_ESTRÉS_Riesgo muy alto,Riesgo psicosocial_AÑO 2025_ESTRÉS_Riesgo alto,Riesgo psicosocial_AÑO 2025_ESTRÉS_Riesgo medio,Riesgo psicosocial_AÑO 2025_ESTRÉS_Riesgo bajo,Riesgo psicosocial_AÑO 2025_ESTRÉS_Sin riesgo,Riesgo psicosocial_AÑO 2025_INTRALABORALES_Riesgo muy alto,Riesgo psicosocial_AÑO 2025_INTRALABORALES_Riesgo alto,Riesgo psicosocial_AÑO 2025_INTRALABORALES_Riesgo medio,Riesgo psicosocial_AÑO 2025_INTRALABORALES_Riesgo bajo,Riesgo psicosocial_AÑO 2025_INTRALABORALES_Sin riesgo,Riesgo psicosocial_AÑO 2025_EXTRABORALES_Riesgo muy alto,Riesgo psicosocial_AÑO 2025_EXTRABORALES_Riesgo alto,Riesgo psicosocial_AÑO 2025_EXTRABORALES_Riesgo medio,Riesgo psicosocial_AÑO 2025_EXTRABORALES_Riesgo bajo,Riesgo psicosocial_AÑO 2025_EXTRABORALES_Sin riesgo,Riesgo psicosocial_AÑO 2023_ESTRÉS_Riesgo muy alto,Riesgo psicosocial_AÑO 2023_ESTRÉS_Riesgo alto,Riesgo psicosocial_AÑO 2023_ESTRÉS_Riesgo medio,Riesgo psicosocial_AÑO 2023_ESTRÉS_Riesgo bajo,Riesgo psicosocial_AÑO 2023_ESTRÉS_Sin riesgo,Riesgo psicosocial_AÑO 2023_INTRALABORALES_Riesgo muy alto,Riesgo psicosocial_AÑO 2023_INTRALABORALES_Riesgo alto,Riesgo psicosocial_AÑO 2023_INTRALABORALES_Riesgo medio,Riesgo psicosocial_AÑO 2023_INTRALABORALES_Riesgo bajo,Riesgo psicosocial_AÑO 2023_INTRALABORALES_Sin riesgo,Riesgo psicosocial_AÑO 2023_EXTRABORALES_Riesgo muy alto,Riesgo psicosocial_AÑO 2023_EXTRABORALES_Riesgo alto,Riesgo psicosocial_AÑO 2023_EXTRABORALES_Riesgo medio,Riesgo psicosocial_AÑO 2023_EXTRABORALES_Riesgo bajo,Riesgo psicosocial_AÑO 2023_EXTRABORALES_Sin riesgo,Evaluación del desempeño_AÑO 2025_OBJETIVO,Evaluación del desempeño_AÑO 2025_COMPETENCIA,Evaluación del desempeño_AÑO 2025_RESULTADO,Evaluación del desempeño_AÑO 2025_INTERPRETACION,Evaluación del desempeño_AÑO 2024_OBJETIVO,Evaluación del desempeño_AÑO 2024_COMPETENCIA,Evaluación del desempeño_AÑO 2024_RESULTADO,Evaluación del desempeño_AÑO 2024_INTERPRETACION,Evaluación del desempeño_AÑO 2023_OBJETIVO,Evaluación del desempeño_AÑO 2023_COMPETENCIA,Evaluación del desempeño_AÑO 2023_RESULTADO,Evaluación del desempeño_AÑO 2023_INTERPRETACION,Evaluación del desempeño_AÑO 2022_OBJETIVO,Evaluación del desempeño_AÑO 2022_COMPETENCIA,Evaluación del desempeño_AÑO 2022_RESULTADO,Evaluación del desempeño_AÑO 2022_INTERPRETACION,Evaluación del desempeño_AÑO 2021_OBJETIVO,Evaluación del desempeño_AÑO 2021_COMPETENCIA,Evaluación del desempeño_AÑO 2021_RESULTADO,Evaluación del desempeño_AÑO 2021_INTERPRETACION,Evaluación del desempeño_AÑO 2020_OBJETIVO,Evaluación del desempeño_AÑO 2020_COMPETENCIA,Evaluación del desempeño_AÑO 2020_RESULTADO,Evaluación del desempeño_AÑO 2020_INTERPRETACION,Engagement_AÑO 2025_AMBIENTE LABORAL,Engagement_AÑO 2025_COMPROMISO,Engagement_AÑO 2025_INTERPRETACION
1,CAUCA,A411243,PDA,2008-04-01,0.222222,0.222222,0.333333,0.222222,0.000000,0.222222,0.222222,0.333333,0.111111,0.111111,0.111111,0.222222,0.555556,0.111111,0.000000,0.363636,0.181818,0.000000,0.272727,0.181818,0.090909,0.090909,0.181818,0.454545,0.181818,0.181818,0.090909,0.363636,0.363636,0.000000,0.807467,0.847273,0.819409,Alto,0.623636,0.839141,0.694909,Medio,0.724494,0.864800,0.765886,Alto,0.671466,0.809636,0.712917,Medio,0.854167,0.865509,0.859838,Alto,0.841818,0.864646,0.853232,Alto,0.840530,0.867424,Medio
2,CAUCA,A411242,PDA,2009-08-01,0.444444,0.333333,0.222222,0.000000,0.000000,0.222222,0.222222,0.111111,0.444444,0.000000,0.111111,0.333333,0.333333,0.000000,0.222222,0.416667,0.166667,0.166667,0.250000,0.000000,0.083333,0.333333,0.333333,0.166667,0.083333,0.250000,0.250000,0.083333,0.166667,0.250000,0.835373,0.865455,0.844397,Alto,0.676417,0.833333,0.723842,Medio,0.707672,0.855636,0.751816,Alto,0.856690,0.788167,0.836133,Alto,0.832500,0.833912,0.833206,Alto,0.943333,0.934606,0.938970,Muy alto,0.843403,0.892361,Medio
3,CAUCA,A411221,PDA,2019-07-15,0.333333,0.166667,0.500000,0.000000,0.000000,0.000000,0.000000,0.666667,0.166667,0.166667,0.000000,0.666667,0.333333,0.00

In [12]:
# Validar el numero de filas y columnas del DataFrame
df_modelo.shape

(186, 61)

**Resumen:** Se identificaron valores faltantes en algunas variables, principalmente asociados a unidades que no cuentan con determinadas mediciones. Para garantizar la comparabilidad de las unidades y reducir el impacto de la ausencia de información, se definió como población de análisis las **186 unidades operativas correspondientes a 121 Agencias y 65 PDA**. Estas unidades serán utilizadas posteriormente en el análisis exploratorio y en el modelo de clusterización.

### 5.4 Análisis de variables categóricas y numéricas

Se realiza una caracterización general de las variables del conjunto de datos, diferenciando entre variables categóricas y numéricas. Este análisis permite conocer la distribución de las principales variables y establecer una base para la selección de los indicadores que posteriormente serán utilizados en el modelo de clusterización.

In [15]:
# Identificar variables categóricas y numéricas
variables_categoricas = df_modelo.select_dtypes(include=['str', 'object', 'category']).columns.tolist()
variables_numericas = df_modelo.select_dtypes(include=['number']).columns.tolist()

print(f"Variables categóricas: {len(variables_categoricas)}")
print(f"Variables numéricas: {len(variables_numericas)}")

print ("----"*20)

print("Variables categóricas:")
for variable in variables_categoricas:
    print(f"- {variable}")

print("\nVariables numéricas:")
for variable in variables_numericas:
    print(f"- {variable}")

Variables categóricas: 10
Variables numéricas: 50
--------------------------------------------------------------------------------
Variables categóricas:
- REGIONAL
- CODIGO UNIDAD
- TIPO
- Evaluación del desempeño_AÑO 2025_INTERPRETACION
- Evaluación del desempeño_AÑO 2024_INTERPRETACION
- Evaluación del desempeño_AÑO 2023_INTERPRETACION
- Evaluación del desempeño_AÑO 2022_INTERPRETACION
- Evaluación del desempeño_AÑO 2021_INTERPRETACION
- Evaluación del desempeño_AÑO 2020_INTERPRETACION
- Engagement_AÑO 2025_INTERPRETACION

Variables numéricas:
- Riesgo psicosocial_AÑO 2025_ESTRÉS_Riesgo muy alto
- Riesgo psicosocial_AÑO 2025_ESTRÉS_Riesgo alto
- Riesgo psicosocial_AÑO 2025_ESTRÉS_Riesgo medio
- Riesgo psicosocial_AÑO 2025_ESTRÉS_Riesgo bajo
- Riesgo psicosocial_AÑO 2025_ESTRÉS_Sin riesgo
- Riesgo psicosocial_AÑO 2025_INTRALABORALES_Riesgo muy alto
- Riesgo psicosocial_AÑO 2025_INTRALABORALES_Riesgo alto
- Riesgo psicosocial_AÑO 2025_INTRALABORALES_Riesgo medio
- Riesgo psicosocial_A

In [16]:
# Resumen de variables numéricas
resumen_numericas = pd.DataFrame({
    'Variable': variables_numericas,
    'Mínimo': df[variables_numericas].min().values,
    'Máximo': df[variables_numericas].max().values,
    'Media': df[variables_numericas].mean().values,
    'Mediana': df[variables_numericas].median().values,
    'Valores faltantes': df[variables_numericas].isna().sum().values
})

resumen_numericas

,Variable,Mínimo,Máximo,Media,Mediana,Valores faltantes
0,Riesgo psicosocial_AÑO 2025_ESTRÉS_Riesgo muy ...,0.000000,0.750000,0.294732,0.285714,52
1,Riesgo psicosocial_AÑO 2025_ESTRÉS_Riesgo alto,0.000000,0.500000,0.195748,0.200000,52
2,Riesgo psicosocial_AÑO 2025_ESTRÉS_Riesgo medio,0.000000,0.600000,0.167173,0.157895,52
3,Riesgo psicosocial_AÑO 2025_ESTRÉS_Riesgo bajo,0.000000,0.666667,0.180408,0.166667,52
4,Riesgo psicosocial_AÑO 2025_ESTRÉS_Sin riesgo,0.000000,0.615385,0.161939,0.142857,52
5,Riesgo psicosocial_AÑO 2025_INTRALABORALES_Rie...,0.000000,0.625000,0.154891,0.142857,52
6,Riesgo psicosocial_AÑO 2025_INTRALABORALES_Rie...,0.000000,0.555556,0.180060,0.173913,52
7,Riesgo psicosocial_AÑO 2025_INTRALABORALES_Rie...,0.000000,0.750000,0.202253,0.200000,52
8,Riesgo psicosocial_AÑO 2025_INTRALABORALES_Rie...,0.000000,0.666667,0.207261,0.200000,52
9,Riesgo psicosocial_AÑO 2025_INTRALABORALES_Sin...,0.000000,1.000000,0.255534,0.214286,52
